source/common/upstream/load_balancer_impl.cc

```diff
@@ -586,10 +591,14 @@ ZoneAwareLoadBalancerBase::hostSourceToUse(LoadBalancerContext* context) {
 
   if (isGlobalPanic(localHostSet())) {
     stats_.lb_local_cluster_not_ok_.inc();
-    // If the local Envoy instances are in global panic, do not do locality
-    // based routing.
-    hosts_source.source_type_ = sourceType(host_availability);
-    return hosts_source;
+    // If the local Envoy instances are in global panic, and we should not fail traffic, do
+    // not do locality based routing.
+    if (fail_traffic_on_panic_) {
+      return absl::nullopt;
```

Comment:
Do we have a unit test that covers this case?

Reply:
I suppose not - I'll add one.